### DISCUSSION TITLE
HOWTO using SPI multi display (hardware mode).

### DISCUSSION THREAD

**Discussion by TrDA-hab** [arendst/Tasmota#13161] [Show and tell]
## 0. About using SPI multi display.   
 - [WIKI SPI](https://en.wikipedia.org/wiki/Serial_Peripheral_Interface).  
 - Tasmota [commands](https://tasmota.github.io/docs/Commands/#displays) for controlling displays.  
 - Tasmota [examples](https://tasmota.github.io/docs/Displays/#rule-examples-for-scripting-examples-see-scripting-docs) for displaying information on the display.  
 - Tasmota [use Buttons](https://tasmota.github.io/docs/Buttons-and-Switches/#button).   
 - Tasmota [use Rules](https://tasmota.github.io/docs/Rules/).    
 - Step by step [instructions for E-Paper Display](https://github.com/arendst/Tasmota/discussions/11850).  
 - You can use any SPI display that uses a 3-wire SPI bus. 4-wire SPI bus - not tested, but should work too (only they are supported by Tasmota).
 - You can connect a maximum of 8 displays using the ESP8266, or more using the ESP32.  
 - You can:  
   - display the same information on all displays at once (by switching on all relays);  
   - or display information individually on each display (including the corresponding relay);  
   - or use mirror display.  

## Other projects where multi display is used:
 - Tasmota [I2C multi display (hardware mode)](https://github.com/arendst/Tasmota/discussions/13166).  
 - Tasmota [SPI+I2C multi display (software mode)](https://github.com/arendst/Tasmota/discussions/13222).  
 - [ESP32 and 8 OLEDs arranged in a circle](https://youtu.be/KZTIBoHtouM). 
 - [Multiple displays on the I2C bus](https://www.youtube.com/watch?v=E9FTQyBYwAE).  
 - [8 Multiplexed OLED /w ESP32 and U8g2](https://www.youtube.com/watch?v=aMgIxXwtHbw).  
 - [SPI OLED multiple displays with an Arduino](https://youtu.be/YCkFFtVEEG4).  
 - [Multiple SPI based Displays with TFT_eSPI on an esp32](https://youtu.be/cCgNHIHijhs).  
 - [Multiple Displays with Arduino](https://youtu.be/yef23sJjiU0).  
 - [Multiple OLED screens](https://youtu.be/TOMkXJWdB4w).  
 - [Multiple OLED Displays / Multiplex TC9548](https://youtu.be/Y9OyLMUgoFk).   

## 1. SPI mirror display (minimal configuration).
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/SPI%20multi%20display/4191.jpg)  
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/SPI%20multi%20display/4192.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/SPI%20multi%20display/20210927_164541.jpg) 

## How to use it.  
 - You must add support for display in `my_user_config.h` file, or flash our ESP8266 module with `tasmota-display.bin` file.   
 - Make all the necessary settings for your display to function.  
 - Run commands in the console for test mirror Display:    
  `Backlog DisplayText [z]; DisplayText [f0s2x2y2tS]; DisplayText [f0s2x2y22tS]; DisplayText [f0s2x2y42tS];`  
  `DisplayText [f0s2x2y62tS]; DisplayText [f0s2x2y82tS]; DisplayText [f0s2x2y102tS]; DisplayText [f0s2x2y122tS];`  
  `DisplayText [f0s2x2y142tS]; DisplayText [f0s2x2y162tS]; DisplayText [f0s2x2y182tS]; DisplayText [f0s2x2y202tS];`  
  `DisplayText [f0s2x2y222tS]; DisplayText [f0s2x2y242tS]; DisplayText [f0s2x2y262tS]; DisplayText [f0s2x2y282tS]`  

Wemos Pin|GPIO|Component|SPI display|
:-:|:-:|:-:|:-:
D5|14|EPaper29 CS|CS
D6|12|SSPI MOSI|DIN
D7|13|SSPI SCLK|CKL

## 2. SPI multi display (medium configuration).

![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/SPI%20multi%20display/4151.jpg)  
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/SPI%20multi%20display/4152.jpg)   
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/SPI%20multi%20display/20210917_191722.jpg)   

#### Video of the display`s work:   
https://youtu.be/MSbM2clI2aU   

## How to use it.  
 - You must add support for display in `my_user_config.h` file, or flash our ESP8266 module with `tasmota-display.bin` file.   
 - Make all the necessary settings for your display to function.  
 - Run the command in the console  to run the "Interlock" mode, for the possibility of individual output of information on displays (optional):  
   `SetOption73 ON` // Enable detach buttons from relays.  
   `Interlock 1,2,3` //Group Relay1 and Relay2 and Relay3 in "group 1".  
   `Interlock 1`     //Enable relay interlock mode.  
 - Run the command in the console to modify the interface (optional):   
   `Backlog WebButton1 E-PAPER#1; WebButton2 E-PAPER#2; WebButton3 E-PAPER#3`  
 - Run commands in the console for test Display:    
   `Backlog Power1 1; DisplayText [z]`  // For run display#1.  
   `DisplayText [f0s2x2y2tS]`    // Print text on display#1.  
   `Backlog Power2 1; DisplayText [z]`  // For run display#2.  
   `DisplayText [f0s2x2y2tS]`    // Print text on display#2.   
   `Backlog Power3 1; DisplayText [z]`  // For run display#3.  
   `DisplayText [f0s2x2y2tS]`    // Print text on display#3.   

Wemos Pin|GPIO|Component|SPI display|Сomment|
:-:|:-:|:-:|:-:|:-:
D3|00|Relay_i 1|SS#1|On/Off display#1
D4|02|Relay_i 2|SS#2|On/Off display#2
D5|14|Relay_i 3|SS#3|On/Off display#3
D6|12|SSPI MOSI|DIN|-
D7|13|SSPI SCLK|CKL|-
D8|15|EPaper29 CS|-|virtual pin

## 3. SPI multi display (maximum configuration).

![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/SPI%20multi%20display/4202.jpg)

**Best regards,   
TrDA.**

**Comment by gemu2015** [comment]
nice finding about multiple display support also your i2c solution.

my multi display solution was indeed poorly described 
it can combine any non universal display (e.g. a matrix display) with any universal display  (up to 3 in total)

#define USE_UNIVERSAL_DISPLAY
#define USE_MULTI_DISPLAY

this example assumes a non universal display at slot 1, but of course all displays may be universal
and adds an sh1106 on slot 2
it then displays daytime on sh1106 every second

\>B
  ; load sh1106 driver on boot
  dt [S2/SH1106_dispay.ini:]
  ; clear screen, switch to LCD font; set auto draw
  dt [zf4s1D1]
; switch to display 1
  dt [S1:]
 \>S
  ; switch to display 2
  dt [S2:]
  ; show time
  dt [x20y20t]
  ; switch back to display 1
  dt [S1:]

since being displaytext commands this will also work with rules

**Reply by TrDA-hab** [reply]
1. Thank you for your feedback !!!
2. Today I was busy making additions to my article: HOWTO using I2C multi display. # 13166
3. In the near future I will test your USE_MULTI_DISPLAY mode.

Question # 1: Where can I see a list of supported displays? 
Question # 2: maybe later ....

**Reply by gemu2015** [reply]
all Tasmota displays are supported on slot 1.  on slot 1, 2 and 3 all universal displays as in folder displaydesc

**Reply by TrDA-hab** [reply]
Cool. Thanks. I'll be testing it soon !!! ... it will take some time.

**Reply by TrDA-hab** [reply]
Worked 4 hours today on your project. 
So far, nothing has worked for me (((

![image](https://user-images.githubusercontent.com/56403720/134050381-650fbb36-b796-4b5c-954f-90c335eb1765.png)
![image](https://user-images.githubusercontent.com/56403720/134050511-3e632f5b-b9a7-4870-8b40-20c0a087c276.png)
![20210920_200700](https://user-images.githubusercontent.com/56403720/134051164-5540f9a7-175c-459b-9f54-130fc761b4a9.jpg)
![20210920_205900](https://user-images.githubusercontent.com/56403720/134051179-8057183a-ef44-44d8-9046-321de858e462.jpg)
)

**Reply by gemu2015** [reply]
ok you want to combine an epaper with an ssd1306.
configure the epaper as usual with 3 wire SPI. it should work out of the box
then copy ssd1306_display.ini to the file system.
specify the i2c pins of the ssd1306 as usual with tasmota pin gui
only define the non universal displays in user config 
#define USE_DISPLAY_EPAPER_29
do not define USE_DISPLAY_SSD1306 since this display driver uses universal display
(however i just saw you use a small ssd1306 for which the descriptor is not yet done)
the descriptor ini file has to be modified for that resolution first.
so better use a supported OLED with is 128x64

then copy the script to the scripter window and enable script
\>B
; load ssd1306 driver on boot
dt [S2/SS1306_dispay.ini:]
; clear screen, switch to LCD font; set auto draw
dt [zf4s1D1]
; switch to display 1
dt [S1:]
\>S
; switch to display 2
dt [S2:]
; show time
dt [x20y20t]
; switch back to display 1
dt [S1:]

then watch the console logging.

**Reply by TrDA-hab** [reply]
`1. Ok you want to combine an epaper with an ssd1306.`

Yes. I want to combine:
-  epaper + epaper. //  SPI + SPI
- or epaper + ssd1306 (64x48). //  SPI + I2C
- or LCD1602 + ssd1306 (64x48). //  I2C + I2C
- or LCD1602 + epaper  //  I2C + SPI + I2C

`2. Configure the epaper as usual with 3 wire SPI. It should work out of the box.`

Yes. This works without issue.

`3. Then copy ssd1306_display.ini to the file system.`

Yes. This works without issue.

`4. Specify the i2c pins of the ssd1306 as usual with tasmota pin gui.`

Yes. This works without issue.

`5. Only define the non universal displays in user config`
`#define USE_DISPLAY_EPAPER_29.`

I understood. I will remove #define USE_DISPLAY_EPAPER_29.

`6. Do not define USE_DISPLAY_SSD1306 since this display driver uses universal display.`

I understood. I will remove #define USE_DISPLAY_SSD1306.

`7. However i just saw you use a small ssd1306 for which the descriptor is not yet done).`

I only have ssd1306 (64x48).

`8. The descriptor ini file has to be modified for that resolution first.`

In the "* .ini" file, only the resolution needs to be changed?

`9. So better use a supported OLED with is 128x64.`

Agree. But they have been ordered and have been traveling for a long time.

**Reply by TrDA-hab** [reply]
@gemu2015 
I've been trying to get your project up and running for a few days now. No result. Please watch the video, maybe you will have recommendations for me.

MULTI_DISPLAY + DISPLAY_EPAPER_29
https://youtu.be/7-y7UPEJmFk

**Reply by gemu2015** [reply]
in your video you selected epaper42 not epaper29
the epaper42 needs too much ram for its frame buffer and will need an esp32 to work.
even the epaper29 needs 4,7 k of ram and together with all the universal driver and the oled RAM may not fit into esp8266 ram space.
all bw displays use a RAM frame buffer which may be too much for an esp8266 in multidisplay mode.
all color displays don't use display frame buffer in ESP memory so work better with an esp8266

in mulidisplay mode each display driver is independent of each other and each display maintains  all its variables like fonts and drawing positions and has it own ram frame buffer on bw displays.

**Reply by TrDA-hab** [reply]
@gemu2015 
Oops sory ... it's my fault. I did a lot of experiments and did a lot of takes on the video.
Now I have fixed everything.

https://youtu.be/jGS9YIIYSp4

**Reply by gemu2015** [reply]
what you get is a stack overflow. 
i guess the epaper alone works as usual.  Check how much ram you have available before you test the multi display mode.
i suggest you switch to an esp32 where you won't have memory problems like this or use colors displays.

i have a working system (wemos esp8266) with an ili9341 and an sh1106 OLED.

**Reply by TrDA-hab** [reply]
@gemu2015 
Good. I will be testing esp32 soon.
For information only
![000](https://user-images.githubusercontent.com/56403720/134394191-a7a0349f-9b38-4f15-9b85-1d43662f205d.jpg).

**Reply by gemu2015** [reply]
is this with epaper initialized and working and script loaded but not enabled ?

**Reply by TrDA-hab** [reply]
Yes.

**Reply by TrDA-hab** [reply]
The loaded script doesn't affect anything - I've tested this.

**Reply by TrDA-hab** [reply]
After epaper initialization, I can load the script only when the ESP8266 itself reboots 5 - 8 times and the WEB interface is available.

**Reply by gemu2015** [reply]
after 6 reboots all pins are reset to default, so epaper is the reason for restart ?

**Reply by TrDA-hab** [reply]
It doesn't matter if I loaded the script or not. ESP8266 reboots without stopping immediately after initializing epaper. Next comes the random events ...

**Reply by gemu2015** [reply]
but epaper alone did work for you previously

**Reply by TrDA-hab** [reply]
`1. After 6 reboots all pins are reset to default.`
Yes. But in very rare cases, there are more than 6 reloads (this is very difficult to catch for video).
`2. So epaper is the reason for restart?`
I think you are absolutely right. After a lot of experimentation.

**Reply by TrDA-hab** [reply]
Yes, this is correct. Unless I used the MULTI_DISPLAY + DISPLAY_EPAPER_29 mode.

**Reply by gemu2015** [reply]
but multi_display does nothing if not started by script.

**Reply by TrDA-hab** [reply]
In my last video: https://youtu.be/jGS9YIIYSp4
You can see - I am not running the script.
This is filmed without stopping and without scripting.
Early in my experiments, I ran a script.
Now I have reduced the number of steps and I can not even get to the step of loading a skip.
ESP8266 reboots many times without stopping.

**Reply by gemu2015** [reply]
try to erase the chip completely. something goes wrong here.  you should get epaper initialized without a reboot first.

simply adding multi_display should definitely no harm if not initialized.

**Reply by TrDA-hab** [reply]
I will test it now ...

**Reply by TrDA-hab** [reply]
> erase the chip completely

Perhaps you are right again. "erase the chip completely" - but how can I just do this, now I have tried many options .... but I could not erase the memory chip!?!

**Reply by TrDA-hab** [reply]
OK. I'll be ready to continue tomorrow.

**Reply by TrDA-hab** [reply]
News.
1. ESP8266:
 - full memory cleaning does not help.
 - the problem is that the ESP8266 has very little operative memory.
 - I commented out as much as possible all unnecessary in my_user_config.h
 - and I was almost able to go through all the steps according to your instructions, but the ESP8266 freezes after executing the script and I did not see the result.

2. ESP32:
 - once I was able to see the results of your script. cool!
 - then the poltergeist came ...
 - my ESP32 disables the SPI bus after first displaying the e-papper.
 - I can’t do anything and clearing the memory completely doesn’t help.

3. I will continue when I have good displays. But it won't be very fast.

**Reply by TrDA-hab** [reply]
@gemu2015
News:
1. I fixed the SPI problem.
2. I need a driver for my ssd1306 (64x48).

![20210925_182826](https://user-images.githubusercontent.com/56403720/134776876-a76fc290-d4a3-4ab5-85ad-6e926260e4eb.jpg)

**Reply by TrDA-hab** [reply]
@gemu2015 
I created this !!!
https://github.com/arendst/Tasmota/discussions/13222

**Reply by gemu2015** [reply]
glad it works now.
the driver for sd1306 fills an area of 128x64 pixels. your displays frame is in this larger frame buffer.
if you specify x and y appropriate you can render your content into this frame.

to adapt your smaller display with accurate frame buffer the display descriptor has to be edited.
1. the resolution has to be adapted as in line 1 => :H,SD1306,64,48,1,I2C,3c,*,*,*
2. these 5 hex bytes adjust display frame buffer transfers.
:A,00,10,40,00,00
unfortunately it is a bit complicated to find the right values for these parameters.
since i don't have such a display i may not help here.
anyhow the 128x64 is much more common and i doubt if is worth the effort to adapt this 64x48

**Reply by TrDA-hab** [reply]
@gemu2015 
News:
1. E-Papper(slot 1) + SSD1306(slot 2) = works well, in the script I passed the offset of the coordinates of the image output.

2. LCD1602(slot 1) + SSD1306(slot 2)  = works randomly.
 - it just doesn't work: https://youtu.be/ci1rZiNQj2A
 - my LCD1602 has been converted to a HID device: https://youtu.be/TqfH2lpTB8E
 
3 .  SSD1306(slot 1) + E-Papper(slot 2)  = does not work, there is no way to assign "EPaper29 CS" to the GPIO (it is not in the list).

Q: What displays can I use in slot 1?

**Reply by gemu2015** [reply]
you can use all displays in slot 1 also not renderer based ones.
any additional display must be universal display driver based (renderer based)
since i dont have LCD displays i could not test but i see no reason why your 1602 should not be combined with an OLED.
there is a very strange crash in your video occurring when you start the script. 
comment out the multi_display cmds in the script and see if the crash persists.

you may even combine more epapers .
universal drivers do not use the specific CS like epaper29_CS
either use SPI_CS or edit the descriptor file and put the CS pin in line one.

H,E-PAPER-29,128,296,1,SPI,1,\*,\*,\*,\*,\*,\*,\*,10

a star (*) means use the tasmota gpio definition, else specify the pin
so probably if pin 20 is your cs specify
H,E-PAPER-29,128,296,1,SPI,1,20,\*,\*,\*,\*,\*,\*,10

 

SPI
1. Number (1 = hardware SPI 1, 2 = Hardware SPI 2 (ESP32), 3 = software SPI
2. CS pin
3. CLK pin
4. MOSI pin
5. DC pin
6. Backlight pin
7. RESET pin
8. MISO pin
9. SPI Speed in MHz

**Reply by TrDA-hab** [reply]
For LCD1602(slot 1) + SSD1306(slot 2):
 - I commented out `// # define USE_MULTI_DISPLAY` in file `user_config_override.h` = no crash.

... tomorrow I will continue.

**Reply by gemu2015** [reply]
i assume your crash is caused by an i2c Adress conflict. HIH Sensor has same address as LCD display (0x27)
so disable HIH sensor and try again.
#undef USE_HIH6

i myself disable all unneeded drivers in user config and only enable those actually used.
this is especially useful on ESP8266 devices to save Flash and RAM

**Reply by sfromis** [reply]
What I do with self-compiled builds is to build time disable all included I2C drivers using `I2CDRIVERS_0_31`, `I2CDRIVERS_32_63` and `I2CDRIVERS_64_95`. Then I generate `I2CDriver` commands from my small configuration management system to enable drivers I specifically want on a device. This still allows run time flexibility of having "too many" drivers included, without risk of these causing conflicts. Of course, I could also generate the masks directly, instead of zeroing everything, but I like a common build for varied devices, including sensors not in the precompiled sensor builds, and excluding sensors I do not own.

**Reply by TrDA-hab** [reply]
@gemu2015 
> so disable HIH sensor and try again.
> #undef USE_HIH6


![image](https://user-images.githubusercontent.com/56403720/134889057-a50bebc7-8cb8-4125-ad33-ecdd5757ffaf.png)
I have tried this now. Tasmota finds HIH6 anyway.
`00:00:00.567 I2C: HIH found at 0x27`

**Reply by TrDA-hab** [reply]
@sfromis 
> What I do with self-compiled builds is to build time disable all included I2C drivers using `I2CDRIVERS_0_31`, `I2CDRIVERS_32_63` and `I2CDRIVERS_64_95`.

I tried to do this. But I am getting an error when compiling.

![image](https://user-images.githubusercontent.com/56403720/134889968-328485fb-ee21-48e1-bc53-4ab3e38a93d6.png)
![image](https://user-images.githubusercontent.com/56403720/134890397-2a325e18-2e1e-4369-919a-1dc04014568e.png)

**Reply by TrDA-hab** [reply]
@gemu2015 
**OK. I ended up testing for USE_MULTI_DISPLAY**
*(maybe I'll continue later when I have other compatible displays).*

1. E-Papper (slot 1) + SSD1306 (slot 2) = works well.
2. LCD1602 (slot 1) + SSD1306 (slot 2) = does not work, **always reboot** if "Script enable" and **Save** the settings.
3. SSD1306 (slot 1) + E-Papper (slot 2) = does not work, **always reboot** if "Script enable" and **Save** the settings.
4. E-Papper (slot 1) + E-Papper (slot 2) = does not work, **always reboot** if "Script enable" and **Save** the settings.

**For your information only.**

**Reply by gemu2015** [reply]
i ordered an LCD panel and tested your configuration. 
now i found the bug. attached are 2 files to replace. please try.  if it works for you also, i will make a pr
[Archiv.zip](https://github.com/arendst/Tasmota/files/7260546/Archiv.zip)

**Reply by TrDA-hab** [reply]
I'll do it now !!! Thanks.

**Comment by gemu2015** [comment]
3 displays at once.
LCD 1602 (i2c), OLED SH1106 (i2c), TFT ST7789 (SPI)

![IMG_3882](https://user-images.githubusercontent.com/11647075/135485038-27a4424a-ceab-4fd4-b336-98639d72201f.jpg)

**Reply by TrDA-hab** [reply]
**Yes. It works !!!**
**Many thanks.**

I've tested:
1. E-Papper (slot 1) + SSD1306 (slot 2) = works well.
2. LCD1602 (slot 1) + SSD1306 (slot 2) = works well.
3. SSD1306 (slot 1) + E-Papper (slot 2) = I don't understand how to configure SPI the driver for slot 2 (E-Papper).
4. E-Papper (slot 1) + E-Papper (slot 2) = I don't understand how to configure SPI the driver for slot 2 (E-Papper).

***I will soon be doing more tests with the SPI driver for the E-Papper (slot 2).***

![20210930_192427](https://user-images.githubusercontent.com/56403720/135497325-5a389f37-eef7-43c4-85b5-2928f7bd91fb.jpg)
![20210930_194300](https://user-images.githubusercontent.com/56403720/135497333-bfb50c42-f9a0-4c0e-8d26-245c8ab7366a.jpg)
